### ISSUE TITLE
Consider alternative operational store

### ISSUE THREAD

**Issue by kefahi** [edraj/dmart#383]
Our current operational store solution (made of Redis + RediSearch + RedisJson) is doing a good job; but it suffers from two specific issues:

1. Large index size (15 million entries are consuming 45GB of ram)
2. In-memory only (i.e. if the data does not fit in memory it wont function)

We would like to consider alternative operational store or indexing engines that can be used instead of Redis.

### Features sought

- Open source 
- Support of Arabic stemming
- Real-time indexing (adding / updating / removing documents)
- Schema-full option (schema validation)
- Small foot print on memory and on disk
- Fast query
- Aggregation
- Mature and actively developed
- Replication / Fail-over
-  Security  / access control 

### Possible optionas 

- [Postgresql](https://www.youtube.com/watch?v=VEWXmdjzIpQ)
- [Meilisearch](https://github.com/meilisearch/meilisearch) : Tested a year a go, it produced huge indexes
- [Sonic](https://github.com/valeriansaliou/sonic)
- [ManticoreSearch](https://github.com/manticoresoftware/manticoresearch) 
- [ZincSearch](https://github.com/zincsearch/zincsearch)
- [TerminusDB](https://terminusdb.com/products/terminusdb/)
- [Tantivy](https://github.com/quickwit-oss/tantivy)

**Comment by kefahi** [comment]
In case we go with RDBMS, here is a possible ERD 
![erd-dmart](https://github.com/edraj/dmart/assets/2543377/e1b203db-c5d4-42a9-9def-514006c927ee)

**Comment by baxiry** [comment]
> Schema-full option (schema validation)

does this mean that it is okay to adopt a relational database?
Or is it okay to abandon JSON as a data storage format?

**Comment by kefahi** [comment]
> > Schema-full option (schema validation)
> 
> does this mean that it is okay to adopt a relational database? Or is it okay to abandon JSON as a data storage format?

Json-based file storage remains as is; as it represents the source of truth. The discussion here is about the ability to index and search content. 
For which we need an "operational" database; which could be document oriented or relational.

**Comment by saadadel** [comment]
I suggest we keep Redis as an optional feature for several reasons

1. Redis memory consumption can be reduced by half if we enhance our implementation, we store two documents for each record
2. Redis has great and complex search queries and aggregation query capabilities
3. We have reached a stable and mature implementation of Redis and doing the same with any other indexing or data storage tool would come with a cost in resources and time
4. After all, we can disable Redis on the entire space by only altering a single space's flag
5. When Redis is not our main database anymore, the two issues described here would disappear

**Comment by saadadel** [comment]
Comparison between the usage of a relational database (with a JSON column for storage and full-text search for querying) and the usage of a non-relational (NoSQL) database:

**Relational Database (with JSON Column and Full-Text Search):**

- Offers structured data storage with JSON documents stored in a single column.
- Full-text search capabilities are available for querying within JSON documents however, it would cause performance overhead in the create and update actions and storage overhead.
- Provides strong data consistency and transaction support (ACID compliance). Which is not essentially needed in our use case.
- Query performance may degrade with large datasets or complex JSON structures.
- Scalability might be limited compared to NoSQL databases, especially for horizontal scaling.
- Backup and restore operations can be resource-intensive, especially with large volumes of data.

**Non-Relational Database (NoSQL):**

- Offers flexible schema design, accommodating dynamic and evolving data structures.
- Optimized for high-speed writes and efficient querying of JSON documents.
- Provides horizontal scalability, making it easier to scale out as the dataset grows.
- Storage overhead is typically minimized, with efficient indexing mechanisms for JSON documents.
- Backup and restore operations are often more efficient compared to relational databases.
- Provides eventual consistency models, trading off strong consistency for scalability and availability.


_In summary, using a non-relational database for storing JSON documents offers advantages in terms of performance, scalability, and flexibility, particularly when dealing with large volumes of data and dynamic data models. NoSQL databases are well-suited for scenarios where the schema is flexible, and there's a need for high-speed reads and writes with efficient querying of JSON documents._

**Comment by baxiry** [comment]
d-mart targets medium-sized businesses and communities with data of up to 300 million entries
This means that we should keep in mind this number declared in the dmart design

Currently, Redis consumes 45 GB of RAM for every 12 million entries
This means 1.2 TB of RAM. 
Suppose the consumption is improved to half.. 20 GB per 10 million entries.
This means 600 GB of RAM to meet promise of 300 million entries.
600 GB of RAM is a very large number and very expensive.

Yes. Demart has reached a good stage of stability with Redis, as you mentioned, and the transformation may require repeated effort and time spent in order to reach this same stage.
But also... the problem presented by Mr. Kifah is a real problem, and may prevent Dmart from spreading on a larger scale in the future.
